### PR TITLE
BUG FIXES: Crash on Exit with partially loaded scene & (near)-infinite loop in skipForward physics

### PR DESCRIPTION
**Pull request by ZappoMan**
BUG FIX 1 - Crash on exit

There was a random (~50% likelihood) crash on exit if you exited the app while some resources were still downloading. The problem was that the "loading" and "pending" resource download requests were being managed in a static member of ResourceCache - but because of C++ static member ctr/dtr management, sometimes those static members are deleted before the users are deleted we'd get a crash.

The fix was to make these members a singleton instead of static members.

Testing:
* open interface - go to apartment
* while content is in process of loading (some entities visible, others not yet visible) Quit the application
* old behavior: crash about 50% of the time
* new behavior: no crash

BUG FIX 2 - Near-infinite loop on skip physics forward

Sometimes the update from the server for changed physics/animated/kinematic objects can result in a dT that is negative or very large. In the very large case we'd end up with what appeared to be a thread deadlock. It was actually a near-infinite loop. We now clamp the dT between 0 and 1 seconds to prevent this.

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -188,6 +189,7 @@ bool setupEssentials(int& argc, char** argv) {
     auto jsConsole = DependencyManager::set<StandAloneJSConsole>();
     auto dialogsManager = DependencyManager::set<DialogsManager>();
     auto bandwidthRecorder = DependencyManager::set<BandwidthRecorder>();
+    auto resouceCacheSharedItems = DependencyManager::set<ResouceCacheSharedItems>();
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: create an instances of our singleton

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -514,6 +516,11 @@ Application::~Application() {
     _myAvatar = NULL;
     
     DependencyManager::destroy<GLCanvas>();
+    DependencyManager::destroy<AnimationCache>();
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: make sure all the ResourceCache derived singletons are deleted before the shared items are

**Comment by hifi-gustavo** [comment]
Build console output for commit: 698fcd81164997dc29eaf892426c826b3e99914e available at: https://gist.github.com/e7f451328083329ec5ef<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4231.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4231.dmg

**Comment by murillodigital** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 698fcd81164997dc29eaf892426c826b3e99914e available at: https://gist.github.com/a0d6720954ab2f8a4963<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4231.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4231.dmg

File: libraries/entities/src/EntityItem.cpp [hunk 3]
```diff
@@ -548,7 +548,10 @@ int EntityItem::readEntityDataFromBuffer(const unsigned char* data, int bytesLef
             // is sending us data with a known "last simulated" time. That time is likely in the past, and therefore
             // this "new" data is actually slightly out of date. We calculate the time we need to skip forward and
             // use our simulation helper routine to get a best estimate of where the entity should be.
-            float skipTimeForward = (float)(now - _lastSimulated) / (float)(USECS_PER_SECOND);
+            const float MIN_TIME_SKIP = 0.0f;
+            const float MAX_TIME_SKIP = 1.0f; // in seconds
+            float skipTimeForward = glm::clamp((float)(now - _lastSimulated) / (float)(USECS_PER_SECOND),
```
Comment (ZappoMan) [inline libraries/entities/src/EntityItem.cpp]: this fixes the near-infinite loop which was sometimes happening.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8fd11e70ccb6cfcf20aeff8c9dbd6d4879896d1e available at: https://gist.github.com/422e0ff74d758781c8a2<br/><br/>Installers available for download at:<br>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4231.exe<br>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4231.dmg

**Pull request merged**